File: Backend-Node/routes/adminReports.js
const express = require('express');
const router = express.Router();
const path = require('path');
const fs = require('fs');
const User = require('../models/User');
const Report = require('../models/Report');

/**
 * @swagger
 * /admin/reports:
 *   get:
 *     summary: Get all reports with user details
 *     tags: [Admin Reports]
 *     responses:
 *       200:
 *         description: Reports retrieved successfully
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 reports:
 *                   type: array
 *                   items:
 *                     type: object
 *                     properties:
 *                       _id:
 *                         type: string
 *                       userName:
 *                         type: string
 *                       userEmail:
 *                         type: string
 *                       reportName:
 *                         type: string
 *                       uploadDate:
 *                         type: string
 *                       type:
 *                         type: string
 *                       totalTests:
 *                         type: number
 *                       abnormalCount:
 *                         type: number
 *                       status:
 *                         type: string
 *                       file_path:
 *                         type: string
 */
router.get('/reports', async (req, res) => {
  try {
    const reports = await Report.find({});
    const result = [];

    for (const report of reports) {
      const user = await User.findOne({ email: report.user_email });

      const severity = report.ai_summary?.severity || 'low';
      let uiStatus = 'normal';
      if (severity === 'medium') {
        uiStatus = 'abnormal';
      } else if (severity === 'high') {
        uiStatus = 'critical';
      }

      // Count abnormal tests
      const testResults = report.testResults || [];
      const abnormalCount = testResults.filter(t => t.status !== 'normal').length;

      result.push({
        _id: report._id.toString(),
        userName: user?.name || 'Unknown',
        userEmail: report.user_email,
        reportName: report.file_name,
        uploadDate: report.uploaded_at,
        type: 'Lab Report',
        totalTests: testResults.length,
        abnormalCount: abnormalCount,
        status: uiStatus,
        file_path: report.file_path
      });
    }

    res.status(200).json({ reports: result });
  } catch (error) {
    console.error('Get all reports error:', error);
    res.status(500).json({ error: 'Server error fetching reports' });
  }
});

/**
 * @swagger
 * /admin/reports/{report_id}:
 *   delete:
 *     summary: Delete a report by MongoDB _id
 *     tags: [Admin Reports]
 *     parameters:
 *       - in: path
 *         name: report_id
 *         required: true
 *         schema:
 *           type: string
 *         description: MongoDB ObjectId of the report
 *     responses:
 *       200:
 *         description: Report deleted successfully
 *       404:
 *         description: Report not found
 */
router.delete('/reports/:report_id', async (req, res) => {
  try {
    const { report_id } = req.params;

    const report = await Report.findById(report_id);
    if (!report) {
      return res.status(404).json({ error: 'Report not found' });
    }

    // Delete file from disk if it exists
    const filePath = report.file_path;
    if (filePath && fs.existsSync(filePath)) {
      fs.unlinkSync(filePath);
    }

    await Report.deleteOne({ _id: report_id });

    res.status(200).json({ message: 'Report deleted successfully' });
  } catch (error) {
    console.error('Delete report error:', error);
    res.status(500).json({ error: 'Server error deleting report' });
  }
});

/**
 * @swagger
 * /admin/download/{report_id}:
 *   get:
 *     summary: Download a report file
 *     tags: [Admin Reports]
 *     parameters:
 *       - in: path
 *         name: report_id
 *         required: true
 *         schema:
 *           type: string
 *         description: MongoDB ObjectId of the report
 *     responses:
 *       200:
 *         description: File downloaded successfully
 *         content:
 *           application/octet-stream:
 *             schema:
 *               type: string
 *               format: binary
 *       404:
 *         description: Report not found
 */
router.get('/download/:report_id', async (req, res) => {
  try {
    const { report_id } = req.params;

    const report = await Report.findById(report_id);
    if (!report) {
      return res.status(404).json({ error: 'Report not found' });
    }

    const filePath = report.file_path;
    if (!filePath || !fs.existsSync(filePath)) {
      return res.status(404).json({ error: 'File not found on server' });
    }

    res.download(filePath, report.file_name);
  } catch (error) {
    console.error('Download report error:', error);
    res.status(500).json({ error: 'Server error downloading report' });
  }
});

/**
 * @swagger
 * /admin/view/{report_id}:
 *   get:
 *     summary: View a report file inline
 *     tags: [Admin Reports]
 *     parameters:
 *       - in: path
 *         name: report_id
 *         required: true
 *         schema:
 *           type: string
 *         description: MongoDB ObjectId of the report
 *     responses:
 *       200:
 *         description: File served successfully
 *       404:
 *         description: Report not found
 */
router.get('/view/:report_id', async (req, res) => {
  try {
    const { report_id } = req.params;

    const report = await Report.findById(report_id);
    if (!report) {
      return res.status(404).json({ error: 'Report not found' });
    }

    const filePath = report.file_path;
    if (!filePath || !fs.existsSync(filePath)) {
      return res.status(404).json({ error: 'File not found on server' });
    }

    res.sendFile(path.resolve(filePath));
  } catch (error) {
    console.error('View report error:', error);
    res.status(500).json({ error: 'Server error viewing report' });
  }
});

module.exports = router;

File: Backend/routes/admin_reports.py
from flask import Blueprint, jsonify, send_file
from pymongo import MongoClient
from bson import ObjectId
import os

mongo = MongoClient("mongodb://localhost:27017/")
db = mongo["LabInsight"]

users_col = db["users"]
reports_col = db["reports"]

admin_reports_bp = Blueprint("admin_reports", __name__)

# ------------------------------------------------
# 1️⃣ GET ALL REPORTS FOR ADMIN PANEL
# ------------------------------------------------
@admin_reports_bp.route("/reports", methods=["GET"])
def get_all_reports():
    reports = list(reports_col.find({}))

    result = []
    for r in reports:
        user = users_col.find_one({"email": r["user_email"]})

        # Map AI severity → UI status
        severity = r.get("ai_summary", {}).get("severity", "low")
        if severity == "low":
            ui_status = "normal"
        elif severity == "medium":
            ui_status = "abnormal"
        else:
            ui_status = "critical"

        result.append({
            "_id": str(r["_id"]),
            "userName": user["name"] if user else "Unknown",
            "userEmail": r["user_email"],
            "reportName": r["file_name"],
            "uploadDate": r["uploaded_at"],
            "type": "Lab Report",
            "totalTests": len(r.get("testResults", [])),
            "abnormalCount": len([t for t in r.get("testResults", []) if t["status"] != "normal"]),
            "status": ui_status,
            "file_path": r["file_path"]
        })

    return jsonify({"reports": result})


# ------------------------------------------------
# 2️⃣ DELETE REPORT
# ------------------------------------------------
@admin_reports_bp.route("/reports/<report_id>", methods=["DELETE"])
def delete_report(report_id):
    report = reports_col.find_one({"_id": ObjectId(report_id)})

    if not report:
        return jsonify({"error": "Report not found"}), 404

    # delete PDF file
    file_path = report.get("file_path")
    if file_path and os.path.exists(file_path):
        os.remove(file_path)

    # delete record
    reports_col.delete_one({"_id": ObjectId(report_id)})

    return jsonify({"message": "Report deleted successfully"})


# ------------------------------------------------
# 3️⃣ DOWNLOAD PDF
# ------------------------------------------------
@admin_reports_bp.route("/download/<report_id>", methods=["GET"])
def download_report(report_id):
    r = reports_col.find_one({"_id": ObjectId(report_id)})

    if not r:
        return jsonify({"error": "Report not found"}), 404

    return send_file(r["file_path"], as_attachment=True)


# ------------------------------------------------
# 4️⃣ VIEW PDF (open in browser)
# ------------------------------------------------
@admin_reports_bp.route("/view/<report_id>", methods=["GET"])
def view_report(report_id):
    r = reports_col.find_one({"_id": ObjectId(report_id)})

    if not r:
        return jsonify({"error": "Report not found"}), 404

    return send_file(r["file_path"])


# ------------------------------------------------
# 5️⃣ DASHBOARD SUMMARY DATA
# ------------------------------------------------

# @admin_reports_bp.route("/dashboard/stats", methods=["GET"])
# def dashboard_stats():
#     total_users = users_col.count_documents({})
#     total_reports = reports_col.count_documents({})

#     # Example: count reports uploaded today
#     from datetime import datetime, timedelta
#     today = datetime.utcnow().date()
#     active_today = reports_col.count_documents({
#         "uploaded_at": {"$gte": str(today)}
#     })

#     avg_reports_per_user = round(total_reports / total_users, 2) if total_users > 0 else 0

#     return jsonify({
#         "total_users": total_users,
#         "total_reports": total_reports,
#         "active_today": active_today,
#         "avg_reports_per_user": avg_reports_per_user
#     })
# @admin_reports_bp.route("/dashboard/activity", methods=["GET"])
# def dashboard_activity():
#     recent = list(
#         reports_col.find({}, {"file_name":1, "user_email":1, "uploaded_at":1})
#                     .sort("_id", -1)
#                     .limit(5)
#     )

#     result = []
#     for r in recent:
#         user = users_col.find_one({"email": r["user_email"]})
#         result.append({
#             "user": user["name"] if user else "Unknown",
#             "action": f"uploaded {r['file_name']}",
#             "time": "just now"   # optional: convert timestamp
#         })

#     return jsonify({"activity": result})
# @admin_reports_bp.route("/dashboard/growth", methods=["GET"])
# def dashboard_growth():
#     pipeline = [
#         {"$group": {
#             "_id": {"$substr": ["$uploaded_at", 0, 7]},  # YYYY-MM
#             "count": {"$sum": 1}
#         }},
#         {"$sort": {"_id": 1}}
#     ]

#     data = list(reports_col.aggregate(pipeline))

#     result = [
#         {"month": d["_id"], "users": d["count"]}
#         for d in data
#     ]

#     return jsonify({"growth": result})
# @admin_reports_bp.route("/dashboard/test-types", methods=["GET"])
# def dashboard_test_types():
#     pipeline = [
#         {"$group": {
#             "_id": "$file_name",
#             "count": {"$sum": 1}
#         }},
#         {"$sort": {"count": -1}},
#         {"$limit": 5}
#     ]

#     data = list(reports_col.aggregate(pipeline))

#     return jsonify({
#         "types": [
#             {"type": d["_id"], "count": d["count"]}
#             for d in data
#         ]
#     })

